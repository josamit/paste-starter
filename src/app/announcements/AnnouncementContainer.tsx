import React, { FC } from "react";
import {
  announcementsNamespacedLSKey,
  DISMISSED,
  useAnnouncement,
} from "@app/announcements/hooks/useAnnouncement";
import { Announcement } from "./types/announcement";
import AnnouncementAlert from "./AnnouncementAlert";

export type AnnouncementProps = Announcement;

const AnnouncementContainer: FC<AnnouncementProps> = ({
  valid_from,
  valid_until,
  variant,
  id,
  content,
  can_be_dismissed,
}) => {
  const [isVisible, setVisible] = useAnnouncement(id, valid_from, valid_until);

  const dismissNotification = () => {
    localStorage.setItem(announcementsNamespacedLSKey(id), DISMISSED);
    setVisible(false);
  };

  return (
    <AnnouncementAlert
      id={id}
      variant={variant}
      content={content}
      isVisible={isVisible}
      onDismiss={can_be_dismissed ? dismissNotification : undefined}
    />
  );
};

export default AnnouncementContainer;
