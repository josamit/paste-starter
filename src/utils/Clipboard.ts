import { UseToasterReturnedProps } from "@twilio-paste/core";

const handleErr = (
  toaster: UseToasterReturnedProps,
  itemName: string,
  err: unknown
) => {
  toaster.push({
    message: `Oops, unable to copy ${itemName} to clipboard. ${err}`,
    variant: "error",
  });
};

const handleSuccess = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "success",
  });
};

const handleWarning = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "warning",
  });
};

export function copyToClipboardSync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  const textArea = document.createElement("textarea");
  textArea.value = textToCopy;

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();

  try {
    const copyResult = document.execCommand("copy");
    const message = copyResult
      ? `${itemName} was successfully copied to your clipboard.`
      : `${itemName} could not be copied to your clipboard. Please try again.`;
    if (copyResult) {
      handleSuccess(toaster, message);
    } else {
      handleWarning(toaster, message);
    }
  } catch (err) {
    handleErr(toaster, itemName, err);
  }

  document.body.removeChild(textArea);
}

export function copyToClipboardAsync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  if (!navigator.clipboard) {
    copyToClipboardSync(textToCopy, itemName, toaster);
    return;
  }

  navigator.clipboard
    .writeText(textToCopy)
    .then(() =>
      handleSuccess(
        toaster,
        `${itemName} was successfully copied to your clipboard.`
      )
    )
    .catch((err) => handleErr(toaster, itemName, err));
}
