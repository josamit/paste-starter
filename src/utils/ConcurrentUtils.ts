/*
 * takes an array of items and a function that accepts an item at a time and returns a promise
 * returns a promise that resolves to an array of results
 */
export function mapConcurrent<T>(
  items: string | any[],
  maxConcurrent: number,
  fn: (arg0: T, arg1: number) => Promise<any>
) {
  let index = 0;
  let currentlyInFlightRequests = 0;
  let completedRequests = 0;
  const results = new Array(items.length);
  let stopExecution = false;

  return new Promise(function (resolve, reject) {
    function run() {
      // launch as many as we're allowed to
      while (
        !stopExecution &&
        currentlyInFlightRequests < maxConcurrent &&
        index < items.length
      ) {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        runNext();
      }
      // if all are done, then resolve parent promise with results
      if (completedRequests === items.length) {
        resolve(results);
      }
    }

    function runNext() {
      const i = index;
      // eslint-disable-next-line no-plusplus
      ++currentlyInFlightRequests;
      // eslint-disable-next-line no-plusplus
      fn(items[index], index++).then(
        function (val) {
          // eslint-disable-next-line no-plusplus
          ++completedRequests;
          // eslint-disable-next-line no-plusplus
          --currentlyInFlightRequests;
          results[i] = val;
          run();
        },
        function (err) {
          // set flag so we don't launch any more requests
          stopExecution = true;
          reject(err);
        }
      );
    }

    run();
  });
}
