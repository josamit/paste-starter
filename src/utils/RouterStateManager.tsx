import * as React from "react";
import { useHistory, useLocation } from "react-router-dom";
import {
  QueryManagerContextProvider,
  QueryParams,
} from "@context/QueryManagerContext";
import {
  emptyQueryParams,
  parseQueryString,
  setQueryParams,
} from "./filtering";

const RouterStateManager: React.FC = ({ children }) => {
  const { search } = useLocation();
  const { push } = useHistory();

  const queryParams = parseQueryString(search || "");
  const newSetQueryParams = (currentPath: string, query: QueryParams) =>
    setQueryParams(push, currentPath, query);
  const newEmptyQueryParams = () => emptyQueryParams(push);

  return (
    <QueryManagerContextProvider
      queryParams={queryParams}
      setQueryParams={newSetQueryParams}
      emptyQueryParams={newEmptyQueryParams}
    >
      {children}
    </QueryManagerContextProvider>
  );
};

export default RouterStateManager;
