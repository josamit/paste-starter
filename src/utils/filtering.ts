import qs from "qs";
import { To } from "history";
import { QueryParams } from "../context/QueryManagerContext";
import { partiallyDecodeURI, partiallyEncodeURI } from "./url";

export function emptyQueryParams(push: (to: To, state?: any) => void) {
  push({ search: "" });
}

export function parseQueryParamsFromQueryString(queryString: string) {
  return qs.parse(queryString, {
    ignoreQueryPrefix: true,
    decoder: partiallyDecodeURI,
  });
}

export function parseQueryString(queryString: string): QueryParams {
  return parseQueryParamsFromQueryString(queryString);
}

export function stringifyWithEncoder(params: Record<string, string>) {
  return qs.stringify(params, {
    encoder: partiallyEncodeURI,
  });
}

export function setQueryParams(
  push: (to: To, state?: any) => void,
  pathname: string,
  params: QueryParams
): void {
  push({
    pathname,
    search: qs.stringify(params),
  });
}
