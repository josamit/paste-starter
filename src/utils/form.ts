import { fireInputChange } from "@twilio/react-typed-forms/dist/testing-utils";
import { fireEvent, RenderResult } from "@testing-library/react";

export function createGetForm({ findByTestId, queryByTestId }: RenderResult) {
  return (formId: string) => ({
    async get(testId: string) {
      return findByTestId(`form-${formId}-${testId}`);
    },
    async click(testId: string) {
      const element = await this.get(testId);

      fireEvent.click(element);
    },
    async type(testId: string, value: string) {
      const element = await this.get(testId);

      fireInputChange(element, value);
    },
    async submit() {
      const element = await this.get("submit-button");

      await fireEvent.click(element);
    },
    async expectValid() {
      const elementId = `${formId}-error`;

      expect(queryByTestId(elementId)).toBeNull();
    },
    async expectValue(testId: string, value: string) {
      const element = (await this.get(testId)) as HTMLInputElement;

      expect(element.value).toEqual(value);
    },
    async select(testId: string, value: string) {
      const element = await this.get(testId);

      await fireEvent.change(element, {
        target: {
          value,
        },
      });
    },
  });
}
