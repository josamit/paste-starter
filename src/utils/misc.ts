import { v4 as uuidv4 } from "uuid";

export function joinClassNames(...classNames: (string | undefined)[]): string {
  return classNames.filter(Boolean).join(" ");
}

// tslint:disable-next-line:no-empty
export const noop = () => {};

export function generateUniqueId() {
  return uuidv4();
}

type TypedEntry<T> = [keyof T, T[keyof T]];

export function typedEntries<T extends object>(obj: T): TypedEntry<T>[] {
  return Object.entries(obj) as TypedEntry<T>[];
}

export function typedKeys<T extends {}>(obj: T): (keyof T)[] {
  return Object.keys(obj) as (keyof T)[];
}

export function safelyParseDate(date?: string | null, asDate = false): string {
  if (date === undefined || date === null || date.trim().length === 0) {
    return "-";
  }
  try {
    const parsedDate = new Date(Date.parse(date));
    if (asDate) {
      return parsedDate.toLocaleDateString();
    }

    return parsedDate.toLocaleString();
  } catch (e) {
    return date;
  }
}
