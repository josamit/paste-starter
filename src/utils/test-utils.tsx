import * as React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Theme } from "@twilio-paste/core/theme";
import { MockQueryManagerContext } from "@context/QueryManagerContext";
import { WindowSizeContext } from "@context/WindowSizeProvider";
import { MemoryRouter, Route, Switch } from "react-router-dom";
import { DarkModeContext } from "@context/DarkModeContext";
import ToastsProvider from "@context/ToastsProvider";
import { createGetForm } from "@src/utils/form";
import { Box } from "@twilio-paste/core/dist/box";

export function renderWithFormUtils(...args: Parameters<typeof rtlRender>) {
  const [children, ...restArgs] = args;

  const results = rtlRender(<Box>{children}</Box>, ...restArgs);
  const getForm = createGetForm(results);

  return {
    ...results,
    getForm,
  };
}

const renderComponent = (
  ui: React.ReactElement,
  {
    initialEntry = "/route/1",
    initialRoute = "/route/:appSid",
    queryParams = {
      accountSid: "ACxxx",
      activeTab: "a2pBrandsReconciliation",
    },
    container = undefined,
    mockToggleMode = jest.fn(),
    setQueryParams = jest.fn(),
    ...renderOptions
  } = {}
) => {
  const Wrapper: React.FC = ({ children }) => (
    <MockQueryManagerContext
      value={{
        queryParams,
        setQueryParams,
      }}
    >
      <WindowSizeContext.Provider
        value={{ width: 1400, height: 800, breakpointIndex: 450 }}
      >
        <Theme.Provider theme="default" disableAnimations>
          <DarkModeContext.Provider
            value={{
              theme: "default",
              toggleMode: mockToggleMode,
              componentMounted: true,
            }}
          >
            <ToastsProvider>
              <MemoryRouter initialEntries={[initialEntry]}>
                <Switch>
                  <Route path={initialRoute}>{children}</Route>
                </Switch>
              </MemoryRouter>
            </ToastsProvider>
          </DarkModeContext.Provider>
        </Theme.Provider>
      </WindowSizeContext.Provider>
    </MockQueryManagerContext>
  );

  return renderWithFormUtils(ui, {
    wrapper: Wrapper,
    container,
    ...renderOptions,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { renderComponent };
