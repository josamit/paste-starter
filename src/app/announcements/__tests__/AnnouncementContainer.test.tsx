import * as React from "react";
import { renderComponent } from "@src/utils/test-utils";
import AnnouncementContainer from "@app/announcements/AnnouncementContainer";

describe("<AnnouncementContainer />", () => {
  it("should render correctly", async () => {
    const { container, getByTestId, getByText } = renderComponent(
      <AnnouncementContainer
        content="[link](https://test.url)<em>test</em>some random text"
        variant="warning"
        can_be_dismissed={false}
        id="test-announcement"
        valid_from="2021-01-01T00:00:00Z"
        valid_until="2099-01-01T00:00:00Z"
      />
    );

    getByText("some random text");

    expect(container).toMatchSnapshot();
  });
});
