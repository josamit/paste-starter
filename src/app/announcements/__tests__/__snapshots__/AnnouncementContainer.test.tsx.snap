// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnnouncementContainer /> should render correctly 1`] = `
<div>
  <div
    class="css-ql39e"
  >
    <div
      class="css-roynbj"
      data-paste-element="BOX"
    >
      <div
        class="css-slcqjs"
        data-paste-element="BOX"
        data-testid="test-announcement"
      >
        <div
          class="css-7efx7i"
          data-paste-element="ALERT"
          data-testid="test-announcement"
          role="alert"
        >
          <div
            class="css-1sekzvp"
            data-paste-element="MEDIA_OBJECT"
            display="flex"
          >
            <div
              class="css-1llkq7k"
              data-paste-element="MEDIA_FIGURE"
              display="flex"
            >
              <span
                class="css-q4ysse"
                color="colorTextIconWarning"
                data-paste-element="ALERT_ICON"
                display="block"
              >
                <svg
                  aria-hidden="false"
                  fill="none"
                  height="100%"
                  role="img"
                  viewBox="0 0 20 20"
                  width="100%"
                >
                  <title>
                    warning: 
                  </title>
                  <path
                    clip-rule="evenodd"
                    d="M17.798 14.791L11.316 3.744c-.582-.992-2.05-.992-2.632 0L2.202 14.79C1.625 15.776 2.354 17 3.518 17h12.964c1.164 0 1.893-1.224 1.316-2.209zM9.99 8h.02c.611.007 1.073.592.978 1.244l-.428 2.238c-.04.296-.278.518-.557.518h-.006c-.279 0-.516-.222-.557-.518l-.428-2.238C8.917 8.592 9.386 8 9.99 8zm.01 7a1 1 0 100-2 1 1 0 000 2z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </div>
            <div
              class="css-1i39mnm"
              data-paste-element="MEDIA_BODY"
            >
              <div>
                <a
                  class="css-lhd4qj"
                  color="colorTextLink"
                  data-paste-element="ANCHOR"
                  font-size="inherit"
                  font-weight="inherit"
                  href="https://test.url"
                  rel="noreferrer noopener"
                  target="_self"
                  text-decoration="underline"
                >
                  link
                </a>
                <span>
                  &lt;em&gt;
                </span>
                <span
                  class="css-1n4qplb"
                  color="colorText"
                  data-paste-element="TEXT"
                  font-size="fontSize30"
                >
                  test
                </span>
                <span>
                  &lt;/em&gt;
                </span>
                <span
                  class="css-1n4qplb"
                  color="colorText"
                  data-paste-element="TEXT"
                  font-size="fontSize30"
                >
                  some random text
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
